pkg/operator/ceph/object/spec.go

```diff
@@ -912,3 +922,93 @@ func renderProbe(cfg rgwProbeConfig) (string, error) {
 
 	return writer.String(), nil
 }
+
+func (c *clusterConfig) addDNSNamesToRGWServer() (string, error) {
+	if (c.store.Spec.Hosting == nil) || len(c.store.Spec.Hosting.DNSNames) <= 0 {
+		return "", nil
+	}
+	if !c.clusterInfo.CephVersion.IsAtLeastReef() {
+		return "", errors.New("rgw dns names are supported from ceph v18 onwards")
+	}
+
+	// add default RGW service name to dns names
+	dnsNames := c.store.Spec.Hosting.DNSNames
+	dnsNames = append(dnsNames, domainNameOfService(c.store))
+
+	// add custom endpoints from zone spec if exists
+	if c.store.Spec.Zone.Name != "" {
+		zone, err := c.context.RookClientset.CephV1().CephObjectZones(c.store.Namespace).Get(c.clusterInfo.Context, c.store.Spec.Zone.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		dnsNames = append(dnsNames, zone.Spec.CustomEndpoints...)
+	}
+
+	// validate dns names
+	var hostNames []string
+	for _, dnsName := range dnsNames {
+		hostName, err := GetHostnameFromEndpoint(dnsName)
+		if err != nil {
+			return "", err
+		}
+		hostNames = append(hostNames, hostName)
+	}
```

Comment:
This is a good optimization to check all the dns names so that this doesn't need called for every section above. However, I can imagine that some users could be confused by where the endpoint is coming from. Let's add a little bit more information so that users are aware where the different hostnames are being retrieved from so they can troubleshoot more on their own:

```suggestion
	// validate dns names
	var hostNames []string
	for _, dnsName := range dnsNames {
		hostName, err := GetHostnameFromEndpoint(dnsName)
		if err != nil {
			return "", errors.Wrap(err, "failed to interpret endpoint from one of the following sources: CephObjectStore.spec.hosting.dnsNames, CephObjectStore.spec.gateway.externalEndpoints, CephObjectZone.spec.customEndpoints"
		}
		hostNames = append(hostNames, hostName)
	}
```

This is also assuming that spec.gateway.externalEndpoints is something we need to read from.